File: BSKeyboardControls/BSKeyboardControls.m
//
//  BSKeyboardControls.m
//  Example
//
//  Created by Simon B. Støvring on 11/01/13.
//  Copyright (c) 2013 simonbs. All rights reserved.
//

#import "BSKeyboardControls.h"

@interface BSKeyboardControls ()
@property (nonatomic, strong) UIToolbar *toolbar;
@property (nonatomic, strong) UISegmentedControl *segmentedControl;
@property (nonatomic, strong) UIBarButtonItem *doneButton;
@property (nonatomic, strong) NSMutableArray *mutableFields;
@end

@implementation BSKeyboardControls

#pragma mark -
#pragma mark Lifecycle

- (id)init
{
    return [self initWithFields:nil];
}

- (id)initWithFrame:(CGRect)frame
{
    return [self initWithFields:nil];
}

- (id)initWithFields:(NSArray *)fields
{
    if (self = [super initWithFrame:CGRectMake(0.0f, 0.0f, 320.0f, 44.0f)])
    {
        [self setToolbar:[[UIToolbar alloc] initWithFrame:self.frame]];
        [self.toolbar setBarStyle:UIBarStyleBlackTranslucent];
        [self.toolbar setAutoresizingMask:(UIViewAutoresizingFlexibleLeftMargin | UIViewAutoresizingFlexibleRightMargin | UIViewAutoresizingFlexibleWidth)];
        [self addSubview:self.toolbar];
        
        [self setSegmentedControl:[[UISegmentedControl alloc] initWithItems:@[ NSLocalizedStringFromTable(@"Previous", @"BSKeyboardControls", @"Previous button title."),
                                                                               NSLocalizedStringFromTable(@"Next", @"BSKeyboardControls", @"Next button title.") ]]];
        [self.segmentedControl addTarget:self action:@selector(segmentedControlValueChanged:) forControlEvents:UIControlEventValueChanged];
        [self.segmentedControl setMomentary:YES];
        [self.segmentedControl setSegmentedControlStyle:UISegmentedControlStyleBar];
        [self.segmentedControl setEnabled:NO forSegmentAtIndex:BSKeyboardControlsDirectionPrevious];
        [self.segmentedControl setEnabled:NO forSegmentAtIndex:BSKeyboardControlsDirectionNext];
        
        [self setDoneButton:[[UIBarButtonItem alloc] initWithTitle:NSLocalizedStringFromTable(@"Done", @"BSKeyboardControls", @"Done button title.")
                                                             style:UIBarButtonItemStyleDone
                                                            target:self
                                                            action:@selector(doneButtonPressed:)]];
        
        [self.toolbar setItems:@[ [[UIBarButtonItem alloc] initWithCustomView:self.segmentedControl],
                                  [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace target:nil action:nil],
                                  self.doneButton ]];
        
        [self setFields:fields];
    }
    
    return self;
}

- (void)dealloc
{
    [self setFields:nil];
    [self setMutableFields:nil];
    [self setSegmentedControlTintControl:nil];
    [self setPreviousTitle:nil];
    [self setBarTintColor:nil];
    [self setNextTitle:nil];
    [self setDoneTitle:nil];
    [self setDoneTintColor:nil];
    [self setActiveField:nil];
    [self setToolbar:nil];
    [self setSegmentedControl:nil];
    [self setDoneButton:nil];
}

#pragma mark -
#pragma mark Public Methods

- (void)setActiveField:(id)activeField
{
    if (activeField != _activeField)
    {
        if ([self.fields containsObject:activeField])
        {
            _activeField = activeField;
        
            if (![activeField isFirstResponder])
            {
                [activeField becomeFirstResponder];
            }
        
            [self updateSegmentedControlEnabledStates];
        }
    }
}

- (void)registerFieldsInView:(UIView *)view
{
    [self addFieldsFromView:view];
    
    NSArray *reversed = [[self.fields reverseObjectEnumerator] allObjects];
    [self setFields:[NSMutableArray arrayWithArray:reversed]];
}

- (void)setFields:(NSArray *)fields
{
    [self setMutableFields:[NSArray arrayWithArray:fields]];
    
    for (UIView *field in fields)
    {
        if ([field isKindOfClass:[UITextField class]])
        {
            [(UITextField *)field setInputAccessoryView:self];
        }
        else if ([field isKindOfClass:[UITextView class]])
        {
            [(UITextView *)field setInputAccessoryView:self];
        }
    }
}

- (NSArray *)fields
{
    return self.mutableFields;
}

- (void)setBarStyle:(UIBarStyle)barStyle
{
    if (barStyle != _barStyle)
    {
        [self.toolbar setBarStyle:barStyle];
        
        _barStyle = barStyle;
    }
}

- (void)setBarTintColor:(UIColor *)barTintColor
{
    if (barTintColor != _barTintColor)
    {
        [self.toolbar setTintColor:barTintColor];
        
        _barTintColor = barTintColor;
    }
}

- (void)setSegmentedControlTintControl:(UIColor *)segmentedControlTintControl
{
    if (segmentedControlTintControl != _segmentedControlTintControl)
    {
        [self.segmentedControl setTintColor:segmentedControlTintControl];
        
        _segmentedControlTintControl = segmentedControlTintControl;
    }
}

- (void)setPreviousTitle:(NSString *)previousTitle
{
    if (![previousTitle isEqualToString:_previousTitle])
    {
        [self.segmentedControl setTitle:previousTitle forSegmentAtIndex:BSKeyboardControlsDirectionPrevious];
        
        _previousTitle = previousTitle;
    }
}

- (void)setNextTitle:(NSString *)nextTitle
{
    if (![nextTitle isEqualToString:_nextTitle])
    {
        [self.segmentedControl setTitle:nextTitle forSegmentAtIndex:BSKeyboardControlsDirectionNext];
        
        _nextTitle = nextTitle;
    }
}

- (void)setDoneTitle:(NSString *)doneTitle
{
    if (![doneTitle isEqualToString:_doneTitle])
    {
        [self.doneButton setTitle:doneTitle];
        
        _doneTitle = doneTitle;
    }
}

- (void)setDoneTintColor:(UIColor *)doneTintColor
{
    if (doneTintColor != _doneTintColor)
    {
        [self.doneButton setTintColor:doneTintColor];
        
        _doneTintColor = doneTintColor;
    }
}

#pragma mark -
#pragma mark Private Methods

- (void)segmentedControlValueChanged:(id)sender
{
    switch (self.segmentedControl.selectedSegmentIndex)
    {
        case BSKeyboardControlsDirectionPrevious:
            [self selectPreviousField];
            break;
        case BSKeyboardControlsDirectionNext:
            [self selectNextField];
            break;
        default:
            break;
    }
}

- (void)doneButtonPressed:(id)sender
{
    if ([self.delegate respondsToSelector:@selector(keyboardControlsDonePressed:)])
    {
        [self.delegate keyboardControlsDonePressed:self];
    }
}

- (void)updateSegmentedControlEnabledStates
{
    NSInteger index = [self.fields indexOfObject:self.activeField];
    if (index != NSNotFound)
    {
        [self.segmentedControl setEnabled:(index > 0) forSegmentAtIndex:BSKeyboardControlsDirectionPrevious];
        [self.segmentedControl setEnabled:(index < [self.fields count] - 1) forSegmentAtIndex:BSKeyboardControlsDirectionNext];
    }
}

- (void)selectPreviousField
{
    NSInteger index = [self.fields indexOfObject:self.activeField];
    if (index > 0)
    {
        index -= 1;
        UIView *field = [self.fields objectAtIndex:index];
        [self setActiveField:field];
        
        if ([self.delegate respondsToSelector:@selector(keyboardControls:selectedField:inDirection:)])
        {
            [self.delegate keyboardControls:self selectedField:field inDirection:BSKeyboardControlsDirectionPrevious];
        }
    }
}

- (void)selectNextField
{
    NSInteger index = [self.fields indexOfObject:self.activeField];
    if (index < [self.fields count] - 1)
    {
        index += 1;
        UIView *field = [self.fields objectAtIndex:index];
        [self setActiveField:field];
        
        if ([self.delegate respondsToSelector:@selector(keyboardControls:selectedField:inDirection:)])
        {
            [self.delegate keyboardControls:self selectedField:field inDirection:BSKeyboardControlsDirectionNext];
        }
    }
}

- (void)addFieldsFromView:(UIView *)view
{
    for (UIView *subview in view.subviews)
    {
        if ([subview isKindOfClass:[UITextField class]] || [subview isKindOfClass:[UITextView class]])
        {
            [self.mutableFields addObject:subview];
        }
        else
        {
            [self addFieldsFromView:subview];
        }
    }
}

@end

File: BSKeyboardControls/BSKeyboardControls.swift
//
//  BSKeyboardControls.swift
//  bskeyboardcontrols
//
//  Created by Ivan Milles on 09/02/15.
//  Copyright (c) 2015 Mr Green. All rights reserved.
//

import UIKit

enum BSKeyboardControl: UInt8 {
	case AllZeros = 0b00
	case PreviousNext = 0b01
	case Done = 0b10
	case AllButtons = 0b11		// Needed for pretty weak Swift NSOption support
}

enum BSKeyboardControlsDirection: Int {
	case Previous = 0
	case Next = 1
}

protocol BSKeyboardControlsDelegate {
	func keyboardControls(keyboardControls: BSKeyboardControls, selectedField field: UIView, inDirection direction: BSKeyboardControlsDirection);
	func keyboardControlsDonePressed(keyboardControls: BSKeyboardControls)
}

class BSKeyboardControls: UIView {
	var toolbar: UIToolbar!
	var doneButton: UIBarButtonItem!
	var previousButton: UIBarButtonItem!
	var nextButton: UIBarButtonItem!
	
	var delegate: BSKeyboardControlsDelegate?
	var visibleControls: BSKeyboardControl = .AllZeros {
		didSet {
			updateToolbar()
		}
	}
	var fields: [UITextField] = [] {		// TODO: Support for UITextView too
		didSet {
			installOnFields()
		}
	}
	
	var activeField: UITextField? {
		didSet {
			activeField?.becomeFirstResponder()
			updatePreviousNextEnabledStates()
		}
	}
	
	var barStyle: UIBarStyle {
		get {return toolbar.barStyle}
		set {toolbar.barStyle = newValue}
	}
	var barTintColor: UIColor? {
		get {return toolbar.barTintColor?}
		set {toolbar.barTintColor = newValue}
	}
	var doneTintColor: UIColor? {
		get {return doneButton.tintColor?}
		set {doneButton.tintColor = newValue}
	}
	var doneTitle: String? {
		get {return doneButton.title}
		set {doneButton.title = newValue}
	}
	
	required convenience init(coder aDecoder: NSCoder) {
		self.init(fields: [])
	}
	
	override convenience init(frame: CGRect) {
		self.init(fields: [])
	}
	
	init(fields: [UITextField]) {
		super.init(frame: CGRect(x: 0.0, y: 0.0, width: 320.0, height: 44.0))
		toolbar = UIToolbar(frame: self.frame)
		barStyle = .Default
		toolbar.autoresizingMask = .FlexibleLeftMargin | .FlexibleRightMargin
		addSubview(toolbar)

		previousButton = UIBarButtonItem(image: UIImage(named: "backbutton"), style: .Plain, target: self, action: "selectPreviousField")
		nextButton = UIBarButtonItem(image: UIImage(named: "nextbutton"), style: .Plain, target: self, action: "selectNextField")
		doneButton = UIBarButtonItem(title: "Done", style: .Done, target: self, action: "doneButtonPressed")
		visibleControls = BSKeyboardControl(rawValue: BSKeyboardControl.PreviousNext.rawValue | BSKeyboardControl.Done.rawValue)!
		
		self.fields = fields

		// didSet observers not called from init()
		installOnFields()
		updateToolbar()
	}
	
	func installOnFields() {
		for field in fields {
			field.inputAccessoryView = self
		}
	}
	
	func updateToolbar() {
		toolbar.items = toolbarItems()
	}
	
	func toolbarItems() -> [AnyObject] {
		var outItems = [AnyObject]()

		if visibleControls.rawValue & BSKeyboardControl.PreviousNext.rawValue > 0 {
			outItems.append(previousButton)
			outItems.append(nextButton)
		}
		
		if visibleControls.rawValue & BSKeyboardControl.Done.rawValue > 0 {
			outItems.append(UIBarButtonItem(barButtonSystemItem: .FlexibleSpace, target: nil, action: nil))
			outItems.append(doneButton)
		}
		
		return outItems
	}
	
	func updatePreviousNextEnabledStates() {
		if let index = find(fields, activeField!) {
			previousButton.enabled = (index > 0)
			nextButton.enabled = (index < fields.count - 1)
		}
	}
	
	func selectPreviousField() {
		if let index = find(fields, activeField!) {
			if index > 0 {
				activeField = fields[index - 1]
				delegate?.keyboardControls(self, selectedField: activeField!, inDirection: .Previous)
			}
		}
	}

	func selectNextField() {
		if let index = find(fields, activeField!) {
			if index < fields.count - 1 {
				activeField = fields[index + 1]
				delegate?.keyboardControls(self, selectedField: activeField!, inDirection: .Next)
			}
		}
	}
	
	func doneButtonPressed() {
		delegate?.keyboardControlsDonePressed(self)
	}
}